File: secureportal.py
from flask import Flask, request, redirect, url_for, render_template_string, session, flash, send_from_directory
from flask_sqlalchemy import SQLAlchemy
from werkzeug.security import generate_password_hash, check_password_hash
from werkzeug.utils import secure_filename
import os, re, logging

# Configuration
app = Flask(__name__)
app.config['SECRET_KEY'] = 'replace_with_your_secret_key'
app.config['SQLALCHEMY_DATABASE_URI'] = 'sqlite:///secureportal.db'
app.config['UPLOAD_FOLDER'] = os.path.join(os.path.dirname(__file__), 'uploads')
app.config['MAX_CONTENT_LENGTH'] = 16 * 1024 * 1024  # 16 MB max upload
ALLOWED_EXTENSIONS = {'pdf'}

db = SQLAlchemy(app)
logging.basicConfig(level=logging.INFO)

# Models
class User(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    username = db.Column(db.String(80), unique=True, nullable=False)
    password_hash = db.Column(db.String(128), nullable=False)
    role = db.Column(db.String(10), nullable=False)  # 'user' or 'admin'
    documents = db.relationship('Document', backref='uploader', lazy=True)

class Document(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    filename = db.Column(db.String(200), nullable=False)
    uploader_id = db.Column(db.Integer, db.ForeignKey('user.id'), nullable=False)

# Helper Functions
def allowed_file(filename):
    return '.' in filename and filename.rsplit('.', 1)[1].lower() in ALLOWED_EXTENSIONS

def current_user():
    if 'user_id' in session:
        return User.query.get(session['user_id'])
    return None

def valid_password(password):
    # Must be at least 8 characters, one uppercase, one lowercase, one special character.
    if len(password) < 8:
        return False
    if not re.search(r'[A-Z]', password):
        return False
    if not re.search(r'[a-z]', password):
        return False
    if not re.search(r'[\W_]', password):
        return False
    return True

# Error Handlers - Generic error messages
@app.errorhandler(Exception)
def handle_exception(e):
    logging.error("Exception occurred", exc_info=e)
    return render_template_string('''
        <!doctype html>
        <html lang="en">
        <head><meta charset="utf-8"><title>Error</title></head>
        <body>
            <h2>An error occurred</h2>
            <p>Please try again later.</p>
            <p><a href="{{ url_for('index') }}">Home</a></p>
        </body>
        </html>
    '''), 500

@app.errorhandler(404)
def not_found_error(e):
    return render_template_string('''
        <!doctype html>
        <html lang="en">
        <head><meta charset="utf-8"><title>Not Found</title></head>
        <body>
            <h2>Page not found</h2>
            <p>The page you are looking for does not exist.</p>
            <p><a href="{{ url_for('index') }}">Home</a></p>
        </body>
        </html>
    '''), 404

# Routes

# Home page
@app.route('/')
def index():
    user = current_user()
    if user:
        return render_template_string('''
            <!doctype html>
            <html lang="en">
            <head>
              <meta charset="utf-8">
              <title>SecurePortal Home</title>
            </head>
            <body>
              <h2>Welcome {{ user.username }} ({{ user.role }})</h2>
              <p><a href="{{ url_for('upload') }}">Upload Document</a></p>
              <p><a href="{{ url_for('documents') }}">View Documents</a></p>
              <p><a href="{{ url_for('logout') }}">Logout</a></p>
            </body>
            </html>
        ''', user=user)
    return render_template_string('''
        <!doctype html>
        <html lang="en">
        <head>
          <meta charset="utf-8">
          <title>SecurePortal</title>
        </head>
        <body>
          <h2>SecurePortal</h2>
          <p><a href="{{ url_for('login') }}">Login</a></p>
          <p><a href="{{ url_for('register') }}">Register</a></p>
        </body>
        </html>
    ''')

# Registration route (HTML and JavaScript for client-side validation)
@app.route('/register', methods=['GET', 'POST'])
def register():
    if request.method == 'POST':
        username = request.form.get('username').strip()
        password = request.form.get('password')
        role = request.form.get('role', 'user')
        if not username or not password:
            flash('Username and password required.')
            return redirect(url_for('register'))
        if not valid_password(password):
            flash('Password must be at least 8 characters long, include one uppercase letter, one lowercase letter, and one special character.')
            return redirect(url_for('register'))
        if User.query.filter_by(username=username).first():
            flash('Username already exists.')
            return redirect(url_for('register'))
        password_hash = generate_password_hash(password)
        new_user = User(username=username, password_hash=password_hash, role=role)
        db.session.add(new_user)
        db.session.commit()
        flash('Registration successful. Please login.')
        return redirect(url_for('login'))
    return render_template_string('''
        <!doctype html>
        <html lang="en">
        <head>
          <meta charset="utf-8">
          <title>Register</title>
          <script>
            function validatePassword() {
                var pwd = document.getElementById("password").value;
                var msg = "";
                if(pwd.length < 8){
                  msg = "Password must be at least 8 characters.";
                } else if(!(/[A-Z]/.test(pwd))){
                  msg = "Password must include at least one uppercase letter.";
                } else if(!(/[a-z]/.test(pwd))){
                  msg = "Password must include at least one lowercase letter.";
                } else if(!(/[\W_]/.test(pwd))){
                  msg = "Password must include at least one special character.";
                }
                document.getElementById("pwdMsg").innerText = msg;
                return msg === "";
            }
          </script>
        </head>
        <body>
          <h2>Register</h2>
          <form method="post" onsubmit="return validatePassword();">
              Username: <input type="text" name="username" required><br>
              Password: <input type="password" name="password" id="password" required onkeyup="validatePassword();"><br>
              <span id="pwdMsg" style="color:red;"></span><br>
              Role:
              <select name="role">
                  <option value="user">User</option>
                  <option value="admin">Admin</option>
              </select><br>
              <input type="submit" value="Register">
          </form>
          <p><a href="{{ url_for('index') }}">Home</a></p>
        </body>
        </html>
    ''')

# Login route
@app.route('/login', methods=['GET', 'POST'])
def login():
    if request.method == 'POST':
        username = request.form.get('username').strip()
        password = request.form.get('password')
        user = User.query.filter_by(username=username).first()
        if user and check_password_hash(user.password_hash, password):
            session['user_id'] = user.id
            flash('Logged in successfully.')
            return redirect(url_for('index'))
        flash('Invalid credentials.')
        return redirect(url_for('login'))
    return render_template_string('''
        <!doctype html>
        <html lang="en">
        <head>
          <meta charset="utf-8">
          <title>Login</title>
        </head>
        <body>
          <h2>Login</h2>
          <form method="post">
              Username: <input type="text" name="username" required><br>
              Password: <input type="password" name="password" required><br>
              <input type="submit" value="Login">
          </form>
          <p><a href="{{ url_for('index') }}">Home</a></p>
        </body>
        </html>
    ''')

# Logout route
@app.route('/logout')
def logout():
    session.pop('user_id', None)
    flash('Logged out.')
    return redirect(url_for('index'))

# Document upload route (Only PDF allowed)
@app.route('/upload', methods=['GET', 'POST'])
def upload():
    user = current_user()
    if not user:
        flash('Please log in to upload documents.')
        return redirect(url_for('login'))
    if request.method == 'POST':
        if 'file' not in request.files:
            flash('No file provided.')
            return redirect(request.url)
        file = request.files['file']
        if file.filename == '':
            flash('No selected file.')
            return redirect(request.url)
        if file and allowed_file(file.filename):
            filename = secure_filename(file.filename)
            os.makedirs(app.config['UPLOAD_FOLDER'], exist_ok=True)
            filepath = os.path.join(app.config['UPLOAD_FOLDER'], filename)
            file.save(filepath)
            new_doc = Document(filename=filename, uploader_id=user.id)
            db.session.add(new_doc)
            db.session.commit()
            flash('File successfully uploaded.')
            return redirect(url_for('documents'))
        else:
            flash('Only PDF files are allowed.')
            return redirect(request.url)
    return render_template_string('''
        <!doctype html>
        <html lang="en">
        <head>
          <meta charset="utf-8">
          <title>Upload Document</title>
        </head>
        <body>
          <h2>Upload Document (PDF only)</h2>
          <form method="post" enctype="multipart/form-data">
              <input type="file" name="file" accept="application/pdf" required><br>
              <input type="submit" value="Upload">
          </form>
          <p><a href="{{ url_for('index') }}">Home</a></p>
        </body>
        </html>
    ''')

# Documents viewing route with least privilege access
@app.route('/documents')
def documents():
    user = current_user()
    if not user:
        flash('Please log in to view documents.')
        return redirect(url_for('login'))
    if user.role == "admin":
        docs = Document.query.all()
    else:
        docs = Document.query.filter_by(uploader_id=user.id).all()
    return render_template_string('''
        <!doctype html>
        <html lang="en">
        <head>
          <meta charset="utf-8">
          <title>Documents</title>
        </head>
        <body>
          <h2>Uploaded Documents</h2>
          <ul>
              {% for doc in docs %}
              <li>{{ doc.filename }} - 
                  <a href="{{ url_for('download_file', filename=doc.filename) }}">Download</a>
              </li>
              {% else %}
              <li>No documents found.</li>
              {% endfor %}
          </ul>
          <p><a href="{{ url_for('index') }}">Home</a></p>
        </body>
        </html>
    ''', docs=docs)

# Download route
@app.route('/uploads/<filename>')
def download_file(filename):
    user = current_user()
    if not user:
        flash('Please log in to download files.')
        return redirect(url_for('login'))
    return send_from_directory(app.config['UPLOAD_FOLDER'], filename, as_attachment=True)

if __name__ == "__main__":
    with app.app_context():
        db.create_all()
    app.run(debug=False)

File: secureportal.js
const express = require('express');
const session = require('express-session');
const bodyParser = require('body-parser');
const { Sequelize, DataTypes } = require('sequelize');
const bcrypt = require('bcrypt');
const multer = require('multer');
const path = require('path');
const fs = require('fs');

const app = express();
app.use(bodyParser.urlencoded({ extended: true }));
app.use(session({
    secret: 'replace_with_your_secret_key',
    resave: false,
    saveUninitialized: false
}));

// Serve static files from the "public" folder
app.use(express.static(path.join(__dirname, 'public')));

// Instead of diskStorage, we use memoryStorage so that file data is available in req.file.buffer
const storage = multer.memoryStorage();
const fileFilter = (req, file, cb) => {
    if (path.extname(file.originalname).toLowerCase() === '.pdf') {
        cb(null, true);
    } else {
        cb(new Error('Only PDF files are allowed'), false);
    }
};
const upload = multer({ storage: storage, fileFilter: fileFilter });

// Initialize Sequelize with SQLite
const sequelize = new Sequelize({
    dialect: 'sqlite',
    storage: 'secureportal.db',
    logging: false
});gslkfdjghghhg

// Define Models
const User = sequelize.define('User', {
    username: {
        type: DataTypes.STRING,
        unique: true,
        allowNull: false
    },
    passwordHash: {
        type: DataTypes.STRING,
        allowNull: false
    },
    role: {
        type: DataTypes.STRING,
        allowNull: false  // 'user' or 'admin'
    }
});

const Document = sequelize.define('Document', {
    filename: {
        type: DataTypes.STRING,
        allowNull: false
    },
    uploaderId: {
        type: DataTypes.INTEGER,
        allowNull: false
    },
    fileData: {
        type: DataTypes.BLOB('long'),
        allowNull: false
    }
});

// Helper function: Enforce strong password policy
function validPassword(password) {
    if (password.length < 8) return false;
    if (!/[A-Z]/.test(password)) return false;
    if (!/[a-z]/.test(password)) return false;
    if (!/[\W_]/.test(password)) return false;
    return true;
}

// Middleware to fetch logged-in user
async function currentUser(req, res, next) {
    if (req.session.userId) {
        try {
            req.user = await User.findByPk(req.session.userId);
        } catch (error) {
            req.user = null;
        }
    }
    next();
}
app.use(currentUser);

// Generic error handler middleware
app.use((err, req, res, next) => {
    console.error(err);
    res.status(500).send(`
        <h2>An error occurred</h2>
        <p>Please try again later.</p>
        <p><a href="/">Home</a></p>
    `);
});

// Routes

// Home page
app.get('/', (req, res) => {
    if (req.user) {
        res.send(`
            <h2>Welcome ${req.user.username} (${req.user.role})</h2>
            <p><a href="/upload.html">Upload Document</a></p>
            <p><a href="/documents.html">View Documents</a></p>
            <p><a href="/logout">Logout</a></p>
        `);
    } else {
        res.send(`
            <h2>SecurePortal</h2>
            <p><a href="/login.html">Login</a></p>
            <p><a href="/register.html">Register</a></p>
        `);
    }
});

// Registration Route
app.get('/register', (req, res) => {
    res.redirect('/register.html');
});
app.post('/register', async (req, res) => {
    const { username, password, role } = req.body;
    if (!username || !password) {
        return res.redirect('/register.html');
    }
    if (!validPassword(password)) {
        return res.send('<p>Weak password. Must be at least 8 chars long with one uppercase, one lowercase, and one special character.</p><p><a href="/register.html">Try again</a></p>');
    }
    const existingUser = await User.findOne({ where: { username } });
    if (existingUser) {
        return res.send('<p>Username already exists.</p><p><a href="/register.html">Try again</a></p>');
    }
    const hash = await bcrypt.hash(password, 10);
    await User.create({ username, passwordHash: hash, role });
    res.send('<p>Registration successful. <a href="/login.html">Login</a></p>');
});

// Login Route
app.get('/login', (req, res) => {
    res.redirect('/login.html');
});
app.post('/login', async (req, res) => {
    const { username, password } = req.body;
    const user = await User.findOne({ where: { username } });
    if (user && await bcrypt.compare(password, user.passwordHash)) {
        req.session.userId = user.id;
        return res.redirect('/');
    }
    res.send('<p>Invalid credentials.</p><p><a href="/login.html">Try again</a></p>');
});

// Logout Route
app.get('/logout', (req, res) => {
    req.session.destroy();
    res.redirect('/');
});

// Upload Route (PDF only)
app.get('/upload', (req, res) => {
    if (!req.user) return res.redirect('/login.html');
    res.redirect('/upload.html');
});
app.post('/upload', upload.single('file'), async (req, res) => {
    if (!req.user) return res.redirect('/login.html');
    if (!req.file) return res.send('<p>No file uploaded.</p><p><a href="/upload.html">Try again</a></p>');
    // Save the file data in the database instead of the file system
    await Document.create({ 
        filename: req.file.originalname,
        uploaderId: req.user.id,
        fileData: req.file.buffer 
    });
    res.send('<p>File successfully uploaded.</p><p><a href="/documents.html">View Documents</a></p>');
});

// Documents Viewing Route
app.get('/documents', async (req, res) => {
    if (!req.user) return res.redirect('/login.html');
    let documents;
    if (req.user.role === 'admin') {
        documents = await Document.findAll();
    } else {
        documents = await Document.findAll({ where: { uploaderId: req.user.id } });
    }
    let listItems = '';
    documents.forEach(doc => {
        listItems += `<li>${doc.filename} - <a href="/uploads/${doc.filename}">Download</a></li>`;
    });
    if (!listItems) listItems = '<li>No documents found.</li>';
    res.send(`
        <h2>Uploaded Documents</h2>
        <ul>${listItems}</ul>
        <p><a href="/">Home</a></p>
    `);
});

// Download Route - Retrieve file data from the database
app.get('/uploads/:filename', async (req, res) => {
    if (!req.user) return res.redirect('/login.html');
    const document = await Document.findOne({ where: { filename: req.params.filename } });
    if (!document) {
        return res.send('<p>File not found.</p><p><a href="/documents.html">Back</a></p>');
    }
    res.set('Content-Type', 'application/pdf');
    res.set('Content-Disposition', `attachment; filename="${document.filename}"`);
    res.send(document.fileData);
});

// Start the server and sync database
sequelize.sync().then(() => {
    app.listen(3000, () => {
        console.log('SecurePortal running on http://localhost:3000');
    });
});